### PR TITLE
ADF4377/ADF4378 driver & application

### PR DESCRIPTION
**Pull request by amiclaus**
- add no-OS driver.
- add IIO support.
- add application example

Signed-off-by: Antoniu Miclaus <antoniu.miclaus@analog.com>

### CODE REVIEW

File: drivers/frequency/adf4377/adf4377.c [hunk 1]
```diff
@@ -0,0 +1,669 @@
...
+ * @param reg_addr - The register address.
+ * @param data - Data value to write.
+ * @return Returns SUCCESS in case of success or negative error code otherwise.
+ */
+int32_t adf4377_spi_write(struct adf4377_dev *dev, uint8_t reg_addr,
+			  uint8_t data)
+{
+	uint8_t buff[ADF4377_BUFF_SIZE_BYTES];
+
+	if (dev->spi_desc->bit_order) {
```
Comment (mchindri) [inline drivers/frequency/adf4377/adf4377.c:64]: I think is not ok to access from the driver members from spi_desc.
You can save bit order in adf4377_dev

File: drivers/frequency/adf4377/adf4377.h [hunk 2]
```diff
@@ -0,0 +1,720 @@
...
+#include "spi.h"
+#include "gpio.h"
+#include "util.h"
+
+/******************************************************************************/
+/********************** Macros and Constants Definitions **********************/
+/******************************************************************************/
+
+/* Registers Control Bits */
+#define ADF4377_REG(x)					(x)
```
Comment (mchindri) [inline drivers/frequency/adf4377/adf4377.h:57]: Why not defining the register values with name:
Ex: ADF4377_REG_SCRATCHPED 0xA
?

Comment (mchindri) [reply drivers/frequency/adf4377/adf4377.h:57]: I will be easy to understand the code

File: drivers/frequency/adf4377/adf4377.c [hunk 3]
```diff
@@ -0,0 +1,669 @@
...
+ */
+int32_t adf4377_soft_reset(struct adf4377_dev *dev)
+{
+	int32_t ret;
+	uint16_t timeout = 0xFFFF;
+	uint8_t data;
+
+	ret = adf4377_update(dev, ADF4377_REG(0x00),
+			     ADF4377_SOFT_RESET_MSK | ADF4377_SOFT_RESET_R_MSK,
+			     ADF4377_SOFT_RESET(ADF4377_SOFT_RESET_EN) | ADF4377_SOFT_RESET_R(
```
Comment (mchindri) [inline drivers/frequency/adf4377/adf4377.c:260]: Nitpick: I would break the line after |

File: drivers/frequency/adf4377/adf4377.c [hunk 4]
```diff
@@ -0,0 +1,669 @@
...
+	if (ret < 0)
+		return ret;
+
+	while(timeout--) {
+		ret = adf4377_spi_read(dev, ADF4377_REG(0x00), &data);
+		if (ret < 0)
+			return ret;
+
+		if(!(data & ADF4377_SOFT_RESET(ADF4377_SOFT_RESET_EN)))
+			return adf4377_set_default(dev);
```
Comment (mchindri) [inline drivers/frequency/adf4377/adf4377.c:271]: Doesn't the software reset update already to the default values?

File: drivers/frequency/adf4377/adf4377.c [hunk 5]
```diff
@@ -0,0 +1,669 @@
...
+	uint8_t data;
+
+	ret = adf4377_update(dev, ADF4377_REG(0x00),
+			     ADF4377_SOFT_RESET_MSK | ADF4377_SOFT_RESET_R_MSK,
+			     ADF4377_SOFT_RESET(ADF4377_SOFT_RESET_EN) | ADF4377_SOFT_RESET_R(
+				     ADF4377_SOFT_RESET_EN));
+	if (ret < 0)
+		return ret;
+
+	while(timeout--) {
```
Comment (mchindri) [inline drivers/frequency/adf4377/adf4377.c:265]: Is timout always enough? 
Whouldn't a delay be useful?

File: drivers/frequency/adf4377/adf4377.h (hunk 2 above)
Comment (amiclaus) [reply drivers/frequency/adf4377/adf4377.h:57]: The names of the registers in datasheet are as follows:
image
image
There are now specific namings for them, that's why i used this approach.

File: drivers/frequency/adf4377/adf4377.c (hunk 1 above)
Comment (amiclaus) [reply drivers/frequency/adf4377/adf4377.c:64]: Wouldn't it be redundant to duplicate an attribute that is already available?

File: drivers/frequency/adf4377/adf4377.c (hunk 3 above)
Comment (amiclaus) [reply drivers/frequency/adf4377/adf4377.c:260]: sure

File: drivers/frequency/adf4377/adf4377.c (hunk 4 above)
Comment (amiclaus) [reply drivers/frequency/adf4377/adf4377.c:271]: Nope, the default values must be rewritten at each re-initialization of the device. I know it's a bit unusual, but this is how this part works.

File: drivers/frequency/adf4377/adf4377.c (hunk 5 above)
Comment (amiclaus) [reply drivers/frequency/adf4377/adf4377.c:265]: It's better to not add delays, only if specified in the datasheet, or if really needed.
The software reset function is called each time the device is initialized and it was tested mutiple times with the afferent hardware.
Therefore, I tend to say no delay is needed.

File: projects/adf4377_sdz/src/app/parameters.h [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+#define GPIO_DEVICE_ID				XPAR_PS7_GPIO_0_DEVICE_ID
+
+#define UART_DEVICE_ID				XPAR_XUARTPS_0_DEVICE_ID
+#define UART_IRQ_ID				    XPAR_XUARTPS_1_INTR
+
+#define UART_BAUDRATE	            115200
+
+#define INTC_DEVICE_ID				XPAR_SCUGIC_SINGLE_DEVICE_ID
+
+#define GPIO_OFFSET					32 + 54
```
Comment (scuciurean) [inline projects/adf4377_sdz/src/app/parameters.h:63]: Maybe we should define this for each supported platform

Comment (amiclaus) [reply projects/adf4377_sdz/src/app/parameters.h:63]: Currently there is only one supported platform, as stated in the app_config.h file. I'd add ifdefs if/when we add support for multiple platforms.

**Comment by amiclaus** [comment]
V2:
- reindent `adf4377_update` function call

File: drivers/frequency/adf4377/adf4377.c [hunk 7]
```diff
@@ -0,0 +1,668 @@
...
+
+/**
+ * @brief Update ADF4377 register.
+ * @param dev - The device structure.
+ * @param reg_addr - The register address.
+ * @param mask - Mask for specific register bits to be updated.
+ * @param data - Data read from the device.
+ * @return Returns SUCCESS in case of success or negative error code otherwise.
+ */
+int32_t adf4377_update(struct adf4377_dev *dev, uint8_t reg_addr,
```
Comment (cristiantpop) [inline drivers/frequency/adf4377/adf4377.c:85]: It is more like a `adf4377_spi_write_mask`. I would rename it.

File: drivers/frequency/adf4377/adf4377.c [hunk 8]
```diff
@@ -0,0 +1,668 @@
...
+		     struct adf4377_init_param *init_param)
+{
+	int32_t ret;
+	uint8_t device_id;
+	uint8_t chip_type;
+	struct adf4377_dev *dev;
+
+	dev = (struct adf4377_dev *)calloc(1, sizeof(*dev));
+	if (!dev)
+		return -ENOMEM;
```
Comment (cristiantpop) [inline drivers/frequency/adf4377/adf4377.c:529]: Add extra line

File: drivers/frequency/adf4377/adf4377.c [hunk 9]
```diff
@@ -0,0 +1,668 @@
...
+			}
+		}
+	} else {
+		ret = -ENODEV;
+		goto error_spi;
+	}
+
+	ret = adf4377_setup(dev);
+	if (ret < 0)
+		goto error_spi;
```
Comment (cristiantpop) [inline drivers/frequency/adf4377/adf4377.c:616]: `goto error_gpio_enclk2;`

File: projects/adf4377_sdz/src/app/adf4377_sdz.c [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+
+	ret = adf4377_spi_read(dev, ADF4377_REG(0x49), &f_lock);
+	if (ret != SUCCESS) {
+		pr_err("SPI Read Failed!");
+	}
+
+	if(ADF4377_LOCKED(f_lock))
+		pr_info("Output Frequency Locked!");
+	else
+		pr_err("Output Frequency not Locked!");
```
Comment (cristiantpop) [inline projects/adf4377_sdz/src/app/adf4377_sdz.c:134]: One is info and one is error?

**Comment by amiclaus** [comment]
V2:
- rebase to master branch for latest driver updates.
- return buffer lenght where needed
- add iio getters where needed

File: projects/adf4377_sdz/src/app/adf4377_sdz.c (hunk 10 above)
Comment (amiclaus) [reply projects/adf4377_sdz/src/app/adf4377_sdz.c:134]: yes. pr_info is used to inform the user that the output was generated successfully.

As for the error:
"pr_err - An error condition, often used by drivers to indicate difficulties with the hardware" (Reference: https://elinux.org/Debugging_by_printing)

File: drivers/frequency/adf4377/adf4377.c (hunk 9 above)
Comment (amiclaus) [reply drivers/frequency/adf4377/adf4377.c:616]: this is valid only for `ADF4377`. 
For `ADF4378` is `goto error_gpio_enclk1;`
Will add extra conditions

**Review by amiclaus — COMMENTED** [review]
V3:
- rename `adf4377_update` function to `adf4377_spi_write_mask`
- update conditional jumps
- add extra line

File: drivers/frequency/adf4377/adf4377.c [hunk 11]
```diff
@@ -0,0 +1,674 @@
...
+	ret = adf4377_spi_read(dev, reg_addr, &read_val);
+	if (ret < 0)
+		return ret;
+
+	read_val &= ~mask;
+	read_val |= data;
+
+	ret = adf4377_spi_write(dev, reg_addr, read_val);
+
+	return ret;
```
Comment (adrimbarean) [inline drivers/frequency/adf4377/adf4377.c:100]: `return adf4377_spi_write(dev, reg_addr, read_val);` is easier.

File: drivers/frequency/adf4377/adf4377.c [hunk 12]
```diff
@@ -0,0 +1,674 @@
...
+
+	return SUCCESS;
+}
+
+/**
+ * @brief Set default datasheet register values.
+ * @param dev - The device structure.
+ * @return Returns SUCCESS in case of success or negative error code.
+ */
+static int32_t adf4377_set_default(struct adf4377_dev *dev)
```
Comment (adrimbarean) [inline drivers/frequency/adf4377/adf4377.c:175]: Is this a default configuration or does it write the initial register value to the registers?

**Review by adrimbarean — APPROVED** [review]
Looks good to me. Just some questions or comments.

**Review by amiclaus — COMMENTED** [review]
V4:
- address latest nitpicks

**Review by cristiantpop — APPROVED** [review]

**Review by scuciurean — APPROVED** [review]

**Pull request merged**